File: iOS/AgoraEntScenarios/Common/API/RTMSyncManager/RTMSyncManager/Classes/Collection/IAUICollection.swift
//
//  IAUICollection.swift
//  AUIKitCore
//
//  Created by wushengtao on 2024/1/4.
//

import Foundation

public typealias AUICollectionGetClosure = (NSError?, Any?)-> Void


//(publisher uid, valueCmd, new value of item) -> value[new value of edit item]
public typealias AUICollectionValueWillChangeClosure = (String, String?, [String: Any]) -> [String: Any]?

//(publisher uid, valueCmd, new value)
public typealias AUICollectionAddClosure = (String, String?, [String: Any]) -> NSError?

//(publisher uid, valueCmd, new value, old value of item)
public typealias AUICollectionUpdateClosure = (String, String?, [String: Any], [String: Any]) -> NSError?

//(publisher uid, valueCmd, oldValue)
public typealias AUICollectionRemoveClosure = (String, String?, [String: Any]) -> NSError?

//(publisher uid, valueCmd, old value of item, keys, update value, min, max)
public typealias AUICollectionCalculateClosure = (String, String?, [String: Any], [String], Int, Int, Int) -> NSError?

//(channelName, key, valueCmd, value[will set metadata])->value[can set metadata]
public typealias AUICollectionAttributesWillSetClosure = (String, String, String?, AUIAttributesModel) -> AUIAttributesModel

//(channelName, key, value)
public typealias AUICollectionAttributesDidChangedClosure = (String, String, AUIAttributesModel) -> Void

@objc public class AUIAttributesModel: NSObject {
    private var attributes: Any?
    public required init(list: [[String: Any]]) {
        self.attributes = list
        super.init()
    }
    public required init(map: [String: Any]) {
        self.attributes = map
        super.init()
    }
    
    public func getMap() -> [String: Any]? {
        return attributes as? [String: Any]
    }
    
    public func getList() -> [[String: Any]]? {
        return attributes as? [[String: Any]]
    }
}

@objc public protocol IAUICollection: NSObjectProtocol {
    
    init(channelName: String, observeKey: String, rtmManager: AUIRtmManager) 
    
    /// 对应的节点对象将要被更新，询问是否需要本地增删(例如更新一个节点，需要再次更新最新时间)
    /// - Parameter callback: <#callback description#>
    @objc optional func subsceibeValueWillChange(callback: AUICollectionValueWillChangeClosure?)
    
    /// 订阅即将添加新的节点的事件
    /// - Parameter callback: <#callback description#>
    @objc optional func subscribeWillAdd(callback: AUICollectionAddClosure?)
    
    /// 订阅即将替换某个节点的事件
    /// - Parameter callback: <#callback description#>
    @objc optional func subscribeWillUpdate(callback: AUICollectionUpdateClosure?)
    
    /// 订阅即将合并某个节点的事件回调
    /// - Parameter callback: <#callback description#>
    @objc optional func subscribeWillMerge(callback: AUICollectionUpdateClosure?)
    
    /// 订阅即将删除某个节点的事件回调
    /// - Parameter callback: <#callback description#>
    @objc optional func subscribeWillRemove(callback: AUICollectionRemoveClosure?)
    
    /// 订阅即将计算某个节点的事件回调
    /// - Parameter callback: <#callback description#>
    @objc optional func subscribeWillCalculate(callback: AUICollectionCalculateClosure?)
    
    /// 即将写入meta data，上层是否需要修改
    /// - Parameter callback: <#callback description#>
    func subscribeAttributesWillSet(callback: AUICollectionAttributesWillSetClosure?)
    
    /// 收到的meta data变化
    /// - Parameter callback: <#callback description#>
    func subscribeAttributesDidChanged(callback: AUICollectionAttributesDidChangedClosure?)
    
    /// 查询当前scene节点所有内容
    /// - Parameter callback: <#callback description#>
    func getMetaData(callback: AUICollectionGetClosure?)
    
    /// 获取本地metadata，仲裁者为本地缓存数据（可能比远端数据更新），观众则为真实远端数据
    /// - Parameter attributes: <#attributes description#>
    func getLocalMetaData() -> AUIAttributesModel?
    
    
    /// 更新本地到远端
    func syncLocalMetaData()
}


@objc public protocol IAUIMapCollection: IAUICollection {
    
    /// 添加节点
    /// - Parameters:
    ///   - valueCmd: <#valueCmd description#>
    ///   - value: <#value description#>
    ///   - callback: <#callback description#>
    func addMetaData(valueCmd: String?,
                     value: [String: Any],
                     callback: ((NSError?)->())?)
    
    /// 更新节点
    /// - Parameters:
    ///   - valueCmd: 命令类型
    ///   - value: <#value description#>
    ///   - callback: <#callback description#>
    func updateMetaData(valueCmd: String?,
                        value: [String: Any],
                        callback: ((NSError?)->())?)
    
    /// 合并节点
    /// - Parameters:
    ///   - valueCmd: <#valueCmd description#>
    ///   - value: <#value description#>
    ///   - callback: <#callback description#>
    func mergeMetaData(valueCmd: String?,
                       value: [String: Any],
                       callback: ((NSError?)->())?)
    
    /// 移除
    /// - Parameters:
    ///   - valueCmd: <#value description#>
    ///   - callback: <#callback description#>
    func removeMetaData(valueCmd: String?,
                        callback: ((NSError?)->())?)
    
    /// 增加/减小节点(节点必须是Int)
    /// - Parameters:
    ///   - valueCmd: <#valueCmd description#>
    ///   - key: <#key description#>
    ///   - value: <#value description#>
    ///   - min: <#min description#>
    ///   - max: <#max description#>
    ///   - callback: <#callback description#>
    func calculateMetaData(valueCmd: String?,
                           key: [String],
                           value: Int,
                           min: Int,
                           max: Int,
                           callback: ((NSError?)->())?)
    
    /// 移除整个collection对应的key
    /// - Parameter callback: <#callback description#>
    func cleanMetaData(callback: ((NSError?)->())?)
}


@objc public protocol IAUIListCollection: IAUICollection {
    
    /// 添加节点
    /// - Parameters:
    ///   - valueCmd: <#valueCmd description#>
    ///   - value: <#value description#>
    ///   - filter: 如果原始数据满足该filter，新增失败，为nil则无条件新增
    ///   - callback: <#callback description#>
    func addMetaData(valueCmd: String?,
                     value: [String: Any],
                     filter: [[String: Any]]?,
                     callback: ((NSError?)->())?)
    
    /// 更新节点
    /// - Parameters:
    ///   - valueCmd: 命令类型
    ///   - value: <#value description#>
    ///   - filter: 如果原始数据满足该filter，才会更新成功，为nil则更新全部
    ///   - callback: <#callback description#>
    func updateMetaData(valueCmd: String?,
                        value: [String: Any],
                        filter: [[String: Any]]?,
                        callback: ((NSError?)->())?)
    
    /// 合并节点
    /// - Parameters:
    ///   - valueCmd: <#valueCmd description#>
    ///   - value: <#value description#>
    ///   - filter: 如果原始数据满足该filter，才会合并成功，为nil则合并全部
    ///   - callback: <#callback description#>
    func mergeMetaData(valueCmd: String?,
                       value: [String: Any],
                       filter: [[String: Any]]?,
                       callback: ((NSError?)->())?)
    
    /// 移除
    /// - Parameters:
    ///   - valueCmd: <#value description#>
    ///   - filter: <#value description#>
    ///   - callback: <#callback description#>
    func removeMetaData(valueCmd: String?,
                        filter: [[String: Any]]?,
                        callback: ((NSError?)->())?)
    
    /// 增加/减小节点(节点必须是Int)
    /// - Parameters:
    ///   - valueCmd: <#valueCmd description#>
    ///   - key: <#key description#>
    ///   - value: <#value description#>
    ///   - min: <#min description#>
    ///   - max: <#max description#>
    ///   - filter: <#filter description#>
    ///   - callback: <#callback description#>
    func calculateMetaData(valueCmd: String?,
                           key: [String],
                           value: Int,
                           min: Int,
                           max: Int,
                           filter: [[String: Any]]?,
                           callback: ((NSError?)->())?)
    
    /// 移除整个collection对应的key
    /// - Parameter callback: <#callback description#>
    func cleanMetaData(callback: ((NSError?)->())?)
}

File: Android/RTMSyncManager/src/main/java/io/agora/rtmsyncmanager/service/collection/IAUICollection.kt
package io.agora.rtmsyncmanager.service.collection

/**
 * Interface for handling various collection operations.
 */
interface IAUICollection {

    /**
     * Subscribes to value changes.
     * @param closure The closure to be called when a value changes.
     */
    fun subscribeValueWillChange(closure: ((publisherId: String, valueCmd: String?, value: Map<String, Any>) -> Map<String, Any>?)?)

    /**
     * Subscribes to additions.
     * @param closure The closure to be called when an addition occurs.
     */
    fun subscribeWillAdd(closure: ((publisherId: String, valueCmd: String?, value: Map<String, Any>) -> AUICollectionException?)?)

    /**
     * Subscribes to updates.
     * @param closure The closure to be called when an update occurs.
     */
    fun subscribeWillUpdate(closure: ((publisherId: String, valueCmd: String?, newValue: Map<String, Any>, oldValue: Map<String, Any>) -> AUICollectionException?)?)

    /**
     * Subscribes to merges.
     * @param closure The closure to be called when a merge occurs.
     */
    fun subscribeWillMerge(closure: ((publisherId: String, valueCmd: String?, newValue: Map<String, Any>, oldValue: Map<String, Any>) -> AUICollectionException?)?)

    /**
     * Subscribes to removals.
     * @param closure The closure to be called when a removal occurs.
     */
    fun subscribeWillRemove(closure: ((publisherId: String, valueCmd: String?, value: Map<String, Any>) -> AUICollectionException?)?)

    /**
     * Subscribes to calculations.
     * @param closure The closure to be called when a calculation occurs.
     */
    fun subscribeWillCalculate(closure: ((publisherId: String, valueCmd: String?, value: Map<String, Any>, cKey: List<String>, cValue: Int, cMin: Int, cMax: Int) -> AUICollectionException?)?)

    /**
     * Subscribes to attribute changes.
     * @param closure The closure to be called when an attribute changes.
     */
    fun subscribeAttributesDidChanged(closure: ((channelName: String, observeKey: String, value: AUIAttributesModel) -> Unit)?)

    /**
     * Subscribes to attribute sets.
     * @param closure The closure to be called when an attribute is set.
     */
    fun subscribeAttributesWillSet(closure: ((channelName: String, observeKey: String, valueCmd: String?, value: AUIAttributesModel) -> AUIAttributesModel)?)

    /**
     * Retrieves the metadata.
     * @param callback The callback to be called with the retrieved metadata.
     */
    fun getMetaData(callback: ((error: AUICollectionException?, value: Any?) -> Unit)?)

    /**
     * Retrieves the local metadata.
     * @return The local metadata.
     */
    fun getLocalMetaData(): AUIAttributesModel?

    /**
     * Releases the collection.
     */
    fun release()
}

/**
 * Interface for handling various map collection operations.
 */
interface IAUIMapCollection : IAUICollection {

    /**
     * Updates the metadata.
     * @param valueCmd The command for the update.
     * @param value The new value.
     * @param callback The callback to be called when the update is complete.
     */
    fun updateMetaData(
        valueCmd: String?,
        value: Map<String, Any>,
        callback: ((error: AUICollectionException?) -> Unit)?
    )

    /**
     * Merges the metadata.
     * @param valueCmd The command for the merge.
     * @param value The value to be merged.
     * @param callback The callback to be called when the merge is complete.
     */
    fun mergeMetaData(
        valueCmd: String?,
        value: Map<String, Any>,
        callback: ((error: AUICollectionException?) -> Unit)?
    )

    /**
     * Adds to the metadata.
     * @param valueCmd The command for the addition.
     * @param value The value to be added.
     * @param callback The callback to be called when the addition is complete.
     */
    fun addMetaData(
        valueCmd: String?,
        value: Map<String, Any>,
        callback: ((error: AUICollectionException?) -> Unit)?
    )

    /**
     * Removes from the metadata.
     * @param valueCmd The command for the removal.
     * @param callback The callback to be called when the removal is complete.
     */
    fun removeMetaData(
        valueCmd: String?,
        callback: ((error: AUICollectionException?) -> Unit)?
    )

    /**
     * Calculates the metadata.
     * @param valueCmd The command for the calculation.
     * @param key The key for the calculation.
     * @param value The value for the calculation.
     * @param min The minimum value for the calculation.
     * @param max The maximum value for the calculation.
     * @param callback The callback to be called when the calculation is complete.
     */
    fun calculateMetaData(
        valueCmd: String?,
        key: List<String>,
        value: Int,
        min: Int,
        max: Int,
        callback: ((error: AUICollectionException?) -> Unit)?
    )

    /**
     * Cleans the metadata.
     * @param callback The callback to be called when the cleaning is complete.
     */
    fun cleanMetaData(callback: ((error: AUICollectionException?) -> Unit)?)
}

/**
 * Interface for handling various list collection operations.
 */
interface IAUIListCollection : IAUICollection {

    /**
     * Updates the metadata.
     * @param valueCmd The command for the update.
     * @param value The new value.
     * @param filter The filter for the update.
     * @param callback The callback to be called when the update is complete.
     */
    fun updateMetaData(
        valueCmd: String?,
        value: Map<String, Any>,
        filter: List<Map<String, Any>>? = null,
        callback: ((error: AUICollectionException?) -> Unit)?
    )

    /**
     * Merges the metadata.
     * @param valueCmd The command for the merge.
     * @param value The value to be merged.
     * @param filter The filter for the merge.
     * @param callback The callback to be called when the merge is complete.
     */
    fun mergeMetaData(
        valueCmd: String?,
        value: Map<String, Any>,
        filter: List<Map<String, Any>>? = null,
        callback: ((error: AUICollectionException?) -> Unit)?
    )

    /**
     * Adds to the metadata.
     * @param valueCmd The command for the addition.
     * @param value The value to be added.
     * @param filter The filter for the addition.
     * @param callback The callback to be called when the addition is complete.
     */
    fun addMetaData(
        valueCmd: String?,
        value: Map<String, Any>,
        filter: List<Map<String, Any>>? = null,
        callback: ((error: AUICollectionException?) -> Unit)?
    )

    /**
     * Removes from the metadata.
     * @param valueCmd The command for the removal.
     * @param filter The filter for the removal.
     * @param callback The callback to be called when the removal is complete.
     */
    fun removeMetaData(
        valueCmd: String?,
        filter: List<Map<String, Any>>? = null,
        callback: ((error: AUICollectionException?) -> Unit)?
    )

    /**
     * Calculates the metadata.
     * @param valueCmd The command for the calculation.
     * @param key The key for the calculation.
     * @param value The value for the calculation.
     * @param min The minimum value for the calculation.
     * @param max The maximum value for the calculation.
     * @param filter The filter for the calculation.
     * @param callback The callback to be called when the calculation is complete.
     */
    fun calculateMetaData(
        valueCmd: String?,
        key: List<String>,
        value: Int,
        min: Int,
        max: Int,
        filter: List<Map<String, Any>>? = null,
        callback: ((error: AUICollectionException?) -> Unit)?
    )

    /**
     * Cleans the metadata.
     * @param callback The callback to be called when the cleaning is complete.
     */
    fun cleanMetaData(callback: ((error: AUICollectionException?) -> Unit)?)
}